cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrade.java

```diff
@@ -0,0 +1,195 @@
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.DatasetDefinition;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.app.runtime.ProgramController;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.data2.dataset2.tx.Transactional;
+import co.cask.cdap.internal.app.store.AppMetadataStore;
+import co.cask.cdap.internal.app.store.ApplicationMeta;
+import co.cask.cdap.internal.app.store.DefaultStore;
+import co.cask.cdap.notifications.feeds.service.MDSNotificationFeedStore;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.RunRecord;
+import co.cask.tephra.TransactionExecutor;
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Iterators;
+import com.google.inject.Injector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Iterator;
+
+/**
+ * Upgraded the Meta Data for applications
+ */
+public class MDSUpgrade extends AbstractUpgrade implements Upgrade {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MDSUpgrade.class);
+  private static final String[] OTHERS = {AppMetadataStore.TYPE_STREAM, AppMetadataStore.TYPE_NAMESPACE,
+    MDSNotificationFeedStore.TYPE_NOTIFICATION_FEED};
+
+  private final Transactional<UpgradeTable, Table> appMDS;
+
+  public MDSUpgrade() {
+    this.appMDS = Transactional.of(executorFactory, new Supplier<UpgradeTable>() {
+      @Override
+      public UpgradeTable get() {
+        try {
+          Table table = DatasetsUtil.getOrCreateDataset(namespacedFramework, Id.DatasetInstance.from
+                                                          (Constants.SYSTEM_NAMESPACE_ID, DefaultStore.APP_META_TABLE),
+                                                        "table", DatasetProperties.EMPTY,
+                                                        DatasetDefinition.NO_ARGUMENTS, null);
+          return new UpgradeTable(table);
+        } catch (Exception e) {
+          LOG.error("Failed to access {} table", DefaultStore.APP_META_TABLE, e);
+          throw Throwables.propagate(e);
+        }
+      }
+    });
+  }
+
+  @Override
+  public void upgrade(Injector injector) throws Exception {
+    appMDS.executeUnchecked(new TransactionExecutor.
+      Function<UpgradeTable, Void>() {
+      @Override
+      public Void apply(UpgradeTable configTable) throws Exception {
+        Scanner rows = configTable.table.scan(null, null);
+        Row row;
+        while ((row = rows.next()) != null) {
+          String key = Bytes.toString(row.getRow()).trim();
+          if (key.contains(AppMetadataStore.TYPE_APP_META)) {
```

Comment:
Should use ``MDSKey`` for deserialization, get the string and then compare with equals, I think. Much cleaner and safer. Perhaps create the ``MDSKey`` here and pass it to the handler methods below, instead of creating it in each method. The String contains operations you have here may not be as safe.